graphql_compiler/compiler/cypher_query.py

```diff
@@ -102,6 +105,52 @@ def _make_query_root_cypher_step(query_metadata_table, linked_location, current_
         where_block=where_block, as_block=as_block)
 
 
+def _generate_cypherstep_list_from_ir_blocks(fold_scope_ir_blocks_dict):
+    """Generate CypherStep object list for each FoldScopeLocation's list of IR blocks.
+
+    Args:
+        fold_scope_ir_blocks_dict: dict of FoldScopeLocation ->
+                                   list of IR blocks corresponding to that @fold scope.
+
+    Returns:
+        dict of FoldScopeLocation -> list of CypherStep objects corresponding to that @fold scope.
+    """
+    if not isinstance(fold_scope_ir_blocks_dict, dict):
+        raise TypeError(u'Expected arguments to be passed through a dictionary. Got {} with type '
+                        u'{} instead.'
+                        .format(fold_scope_ir_blocks_dict,
+                                type(fold_scope_ir_blocks_dict).__name__))
+    for fold_scope_location in fold_scope_ir_blocks_dict:
+        folded_cypher_steps = []  # list of CypherStep objects corresponding to this fold scope.
+        linked_location = fold_scope_location.base_location
+        step_block = Fold(fold_scope_location)  # for first CypherStep object in this fold scope.
+        fold_scope_ir_blocks = [step_block] + fold_scope_ir_blocks_dict[fold_scope_location]
+        query_metadata_table = None  # In _make_cypher_step, this is needed for only two
+        # things: making a query root CypherStep and getting all the supertypes of the type
+        # specified in the coercion block. However, this is not a root block because we're
```

Comment:
Any reason not to just make `Fold` be a root block? I'm not sure I understand what's going on here, and I'm concerned about setting `query_metadata_table = None` when we could have possibly just passed in the "real" metadata table object.

Reply:
Right now Python complains about it being unhashable, which means that if we mutate it anywhere, we can't use it as a dictionary key in a CypherQuery's `fold` field. If it's immutable, we could probably just use the Fold block's FoldScopeLocation's hash as its hash?

`_make_cypher_step` requires parameters `query_metadata_table`, `linked_location`, `current_step_blocks`. 

Query metadata table is used for two things: making a QueryRoot block (which doesn't apply) and computing `step_types`. However, `_get_all_supertypes_of_exact_type` doesn't use this parameter anyways, so essentially for working with folds, query metadata table doesn't seem to serve any purpose.

Linked location is where you're coming from in the traversal

Current step blocks is the list of blocks that we'll generate a CypherStep with